crates/core/component/dex/src/component/position_manager.rs

```diff
@@ -187,16 +197,42 @@ pub trait PositionManager: StateWrite + PositionRead {
     /// Combines a list of fixed candidates with a list of liquidity-based candidates.
     /// This ensures that the fixed candidates are always considered, minimizing
     /// the risk of attacks on routing.
-    async fn candidate_set(
+    fn candidate_set(
         &self,
-        _from: asset::Id,
+        from: asset::Id,
         fixed_candidates: Arc<Vec<asset::Id>>,
-    ) -> Result<Vec<asset::Id>> {
-        let candidates = BTreeSet::from_iter(fixed_candidates.iter().cloned());
-        // TODO: do dynamic candidate selection based on liquidity (tracked by #2750)
-        // Idea: each block, compute the per-asset candidate set and store it
-        // in the object store as a BTreeMap.
-        Ok(candidates.into_iter().collect())
+    ) -> Pin<Box<dyn Stream<Item = Result<asset::Id>> + Send>> {
+        let mut dynamic_candidates = self
+            .ordered_routable_assets(&from)
+            .take(DYNAMIC_ASSET_LIMIT);
+        try_stream! {
+            // First stream the fixed candidates, so those can be processed while the dynamic candidates are fetched.
+            for candidate in fixed_candidates.iter() {
+                yield candidate.clone();
+            }
+
+            // Yield the liquidity-based candidates. Note that this _may_ include some assets already included in the fixed set.
+            while let Some(candidate) = dynamic_candidates
+                .next().await {
```

Comment:
Thinking more about it, I think you're right that dupes are going to be more the rule than the exception. Henry suggested using a filter on the stream iterator. That's probably a good idea